BLE Keyboards

bmcgavin opened the pull request:
Hello, love your work. I've got a lot of ZMK keyboards, which use nrf52840 controllers for BLE communication, so I thought I'd make an attempt at adding in BLE support to the MJ (revision 5 - no reason it shouldn't work with 4, but it's already got a keyboard attached?).

To have a go, add your BLE device's address to config.json as ble.remote, e.g.:

```
  "ble": {
    "remote": "xx:xx:xx:xx:xx:xx"
  }
```

Then the MJ will try to connect to this device on boot. There's a new status icon between the locale and the save indicator - grey for disconnected, blue for connected.

Things to note:

* I am not a C programmer so I am copying from tutorials and hacking things in where needed (see the public methods in the EspUsbHost class)
* This blocks startup for a second or two - until the keyboard is connected.
* I initially started coming at this from the menu, so there is a broken menu option that used to initiate a scan, this is probably redundant now that the config.json option exists.
* The keyboard needs to be on when the MJ powers up otherwise it won't find it (this delays startup too).

Still to do:

* Menu options for scanning and saving keyboards maybe?
* Disconnect / reconnect handling
* Bond storage [X]
* Code cleanup
* Any kind of status indicator [X]

Warnings:

* I use Linux so the filesystem is case sensitive - I had to change Animatedgif.h to AnimatedGIF.h in micro-journal-rev-4-esp32/src/display/WakeUp/WakeUp.h - I hope this still works on your build system.
* I initially used the bluedroid-based BLE library, but this required changing the partition table to fit - I have since migrated to NimBLE which is smaller and I _think_ I've reverted the partition size to the defaults?
* I had to comment out a huge amount of USB Host debug logging to make debug logging usable to check BLE errors - I'm reverting these changes for this PR.

I hope you find this interesting or helpful - I'm now going on holiday for a week so will be able to answer questions here but not really do any deep debugging. Thanks for creating the micro journal, this week of soldering, gluing, 3D printing and C development have been really fun (but maybe a bit of a distraction away from writing!)

unkyulee commented:
I can not belive what I am reading right now. 

Thank you so much for sharing your talent. I will review every time with great appreciation and will use it to share with others.
I am so excited to test this feature out!


Thank you!

Un Kyu Lee

bmcgavin commented:
Slight update from the last two weeks:

* Bonding now works - The keyboard will be connected on startup if it's ready to connect when the MJ is turned on.
* Couple of fixes to the key handling to improve accuracy.
* Changes to the config structure - this uses the SD config rather than the SPIFF partition(?) in order to allow the ble data to be read as part of the startup.
* The remote address can now be added via the menu - reboot to connect.

My main issues now are:

* There's no re-connection mechanism - my logic here is that reboots are quick, so if your keyboard disconnected then restart the MJ. I couldn't make reconnection work in a background thread, so it would block any attempts to type on a backup USB keyboard.
* Scanning - several issues here
  * I am bad at working with C++ string arrays, so I'm having some serious troubles getting the flow of data (scan callback -> array of addresses -> UI output) to work.
  * It seems that having notify callbacks (for keypresses) and scan callbacks (for devices found in scanning) causes some threading problems - even with tearing down an existing BLE connection as much as possible, the scan() call would hang.
    *  I got as far as enabling a 'scan mode' config flag to allow the boot sequence to skip device connection, so a USB keyboard can be used to initiate a scan.

I'm tempted to disable scanning for now and just allow people to add the ble.remote data directly.

bmcgavin commented:
Scanning and device selection now works - I was trying to print std::string as char*, that's why I wasn't getting very far with printing the found devices.

Also there was a bug where I'd not disabled the infinite reconnect loop if a bluetooth device is set but not found, so I've fixed that.

I added a restart option from the BLE menu in order to save wear on my ESP plug.

I'm not going to make any substantial feature changes on this now so I'll remove the WIP from the PR title. Happy to help with any bugs you find in testing.

Thanks again for your work, looking forward to the revision 2 revamp as well.

unkyulee commented:
Incredible! 
I will merge this changes in once Rev.2 revamp gets released. 
I can't wait to have your code tested!!

Thanks a lot for sharing your talent

Un Kyu Lee

darryllee commented:
I would love to test this on a Rev.5 if you can build an image I could install. Thank you!

unkyulee commented:
It seems that it only works with BLE keyboards, specifically with ZMK keyboard. 
I need to research a bit better, if this is possible to connect to other types of bluetooth keyboards. 
I am currently lacking too much knowledge to release this version. 
It is already merged in the main codes, so at the next release binary, you should be able to try it out.

bmcgavin commented:
Hello, thanks for the merge!

The code uses the BLE service UUID 1812 which is for Human Interface Devices (i.e. keyboards etc). Within that service it uses characteristic 2a4d which is for 'Report' - these are pretty generic and may work on any BLE keyboard, not just ZMK. However, it's possible that the service or characteristic UUIDs are different for other keyboard firmwares, which should be a quick fix once those UUIDs are known.

As an aside, the service / characteristic UUIDs are all defined by the Bluetooth SIG - https://www.bluetooth.com/specifications/assigned-numbers/

Unfortunately I don't have any BLE keyboards that are not ZMK powered, so I would need your support in connecting to a keyboard and providing service/characteristic UUIDs it advertises. This could be done using an app like LightBlue and providing service UUIDs, and characteristic UUIDs for each of the services your keyboard advertises.

Bluetooth (non-LE) isn't supported by the NimBLE Bluetooth stack. It is supported by Bluedroid but I'm not certain if the esp32s3 hardware supports it. To test this would require changing the partition structure as the library is massive, and I don't have any Bluetooth (non-LE) keyboards to test with either.

darryllee commented:
Hi - so my new keyboard that I bought specifically to use with my rev.5 is a Royal Kludge F68. Here's what my Mac reports for it. Do you think it might be sufficient for pairing using your new feature?
>   RK-F68 5.0 :
>   Address:	CD:BB:23:9C:A6:EF
>   Battery Level:	100%
>   Minor Type:	Keyboard
>   Services:	0x400000 < BLE >

bmcgavin commented:
Service id 0x400000 isn't in the Bluetooth SIG docs, so I think that is not the service UUID that would need to be added to the code.

I haven't used this app and so do not endorse it, but something like this  https://apps.apple.com/gb/app/ble-scanner-4-0/id1221763603 looks like it could help get the service UUID.

darryllee commented:
Hi, so I used this app to scan and it displayed this Device UUID:
> 41349436-7614-DBFB-BEF6-281D5BA3DC6F

Additionally, under **Services** it shows:
> Device Information:
> 0x180A
> Primary Service

And when I expand that it shows me:
> **SOFTWARE REVISION STRING (2A28)**
> **Read**
> Properties
> **0x{length=4,bytes=0x30312e36}** 
> Value - HEX at 03:10:47.538
> **01.6**
> Value - String at 03:10:47.538

> **PNP ID (2A50)**
> **Read**
> Properties
> **0x{length=7,bytes=0x0200012340167}**
> Value - HEX at 03:10:47.569
> Value - String at 03:10:47.569

> **MANUFACTURER NAME STRING (2A29)**
> **Read**
> Properties
> **0x{length=5,bytes=0x42454b454e}**
> Value - HEX at 03:10:47.506
> **BEKEN**
> Value - String at 03:10:47.506

bmcgavin commented:
I think (and this is my first Bluetooth development work so bear with me) that service is just for displaying device information. Are there any other services exposed?

bmcgavin commented:
Just a thought - the keyboard you are using supports both Bluetooth and wireless 2.4GHz (via a dongle?). Please can you confirm that it was in Bluetooth mode when you gathered the data?

darryllee commented:
Hi - actually this keyboard only has Bluetooth and wired USB. No dongle.

Here's screenshots of every screen that is available from the scanner: https://photos.app.goo.gl/VzJAfaUuZabqe1mQ9

It appears to be reporting Service UUIDs as "HumanInterfaceDevice"

bmcgavin commented:
It does appear to be reporting HID service UUIDs, but not showing the 1812 UUID I'm expecting.

This may be an artefact of the app not showing the information I'm looking for. Is there a way of seeing the advertisement packets it intercepts, maybe on the 'advertisement' tab?

It's also possible everything would 'just work' if you had the firmware. @unkyulee are you okay with me providing a release firmware on my fork of this repo, with plenty of warnings about the unsupported nature?
